cpp/dolfinx/io/xdmf_function.cpp

```diff
@@ -88,98 +56,153 @@ void xdmf_function::add_function(MPI_Comm comm, const fem::Function<T, U>& u,
   assert(u.function_space());
   auto mesh = u.function_space()->mesh();
   assert(mesh);
+  auto element = u.function_space()->element();
+  assert(element);
+
+  // FIXME: is the below check adequate for detecting a Lagrange
+  // element?
+  // Check that element is Lagrange
+  if (!element->interpolation_ident())
+  {
+    throw std::runtime_error("Only Lagrange functions are supported. "
+                             "Interpolate Functions before output.");
+  }
+
+  auto map_c = mesh->topology()->index_map(mesh->topology()->dim());
+  assert(map_c);
+
+  auto map_x = mesh->geometry().index_map();
+  assert(map_x);
+
+  auto dofmap = u.function_space()->dofmap();
+  assert(dofmap);
+  const int bs = dofmap->bs();
+
+  int rank = element->value_shape().size();
+  int num_components = std::pow(3, rank);
 
   // Get fem::Function data values and shape
   std::vector<T> data_values;
-  const bool cell_centred = has_cell_centred_data(u);
+  std::span<const T> x = u.x()->array();
+
+  const bool cell_centred
+      = element->space_dimension() / element->block_size() == 1;
   if (cell_centred)
-    data_values = xdmf_utils::get_cell_data_values(u);
+  {
+    // Get dof array and pack into array (padded where appropriate)
+    const std::int32_t num_local_cells = map_c->size_local();
+    data_values.resize(num_local_cells * num_components, 0);
+    for (std::int32_t c = 0; c < num_local_cells; ++c)
+    {
+      auto dofs = dofmap->cell_dofs(c);
+      assert(dofs.size() == 1);
+      for (std::size_t i = 0; i < dofs.size(); ++i)
+        for (int j = 0; j < bs; ++j)
+          data_values[num_components * c + j] = x[bs * dofs[i] + j];
+    }
+  }
   else
-    data_values = xdmf_utils::get_point_data_values(u);
+  {
+    // Get number of geometry nodes per cell
+    const auto& geometry = mesh->geometry();
+    auto& cmap = geometry.cmaps()[0];
+    int cmap_dim = cmap.dim();
+    int cell_dim = element->space_dimension() / element->block_size();
+    if (cmap_dim != cell_dim)
+    {
+      throw std::runtime_error(
+          "Degree of output Function must be same as mesh degree. Maybe the "
+          "Function needs to be interpolated?");
+    }
 
-  auto map_c = mesh->topology()->index_map(mesh->topology()->dim());
-  assert(map_c);
+    // Check that dofmap layouts are equal and check Lagrange variants
+    if (dofmap->element_dof_layout() != cmap.create_dof_layout())
+    {
+      throw std::runtime_error("Function and Mesh dof layouts do not match. "
+                               "Maybe the Function needs to be interpolated?");
+    }
+    if (cmap.degree() > 2
+        and element->basix_element().lagrange_variant() != cmap.variant())
```

Comment:
Do we really need the lagrange varaint check as for degree less than 3 the spacing shouldn't matter?

Reply:
Hence the first check above - `cmap.degree() > 2`.